### PR TITLE
[Datastore] Add S3 and Kafka support to datastore profiles

### PR DESCRIPTION
**Pull request by alxtkr77**

### CODE REVIEW

**Review by gtopper — COMMENTED** [review]
A very partial review for now because apparently this isn't for 1.5.0.

File: mlrun/datastore/s3.py [hunk 1]
```diff
@@ -104,30 +117,47 @@ def get_filesystem(self, silent=False):
                     "AWS s3fs not installed, run pip install s3fs"
                 ) from exc
             return None
-        self._filesystem = fsspec.filesystem("s3", **self.get_storage_options())
+
+        class S3withDS(s3fs.S3FileSystem):
```
[the remark this answers is not in the document]
Comment (alxtkr77) [reply mlrun/datastore/s3.py:121]: It must be here since it depends on 'import s3fs' which is couple of lines above

File: mlrun/datastore/datastore_profile.py [hunk 2]
```diff
@@ -48,6 +49,75 @@ def generate_secret_key(profile_name: str, project: str):
...
+    topic: str
+    kwargs_public: typing.Optional[typing.Dict]
+    kwargs_private: typing.Optional[typing.Dict]
+
+    def attributes(self):
+        result = {"bootstrap_servers": self.bootstrap_servers}
+        if self.kwargs_public:
+            result = merge(result, self.kwargs_public)
+        if self.kwargs_private:
+            result = merge(result, self.kwargs_private)
```
[the remark this answers is not in the document]
Comment (alxtkr77) [reply mlrun/datastore/datastore_profile.py:65]: Frankly I did not give any thought on the priority, since I assumed there are no duplications. What is the order you think should be here?

File: mlrun/datastore/datastore.py [hunk 3]
```diff
@@ -50,7 +50,7 @@ def parse_url(url):
 
 def schema_to_store(schema):
     # import store classes inside to enable making their dependencies optional (package extras)
-    if not schema or schema in ["file"] + list(string.ascii_lowercase):
+    if not schema or schema in ["file", "kafkaTarget"] + list(string.ascii_lowercase):
```
[the remark this answers is not in the document]
Comment (gtopper) [reply mlrun/datastore/datastore.py:54]: I'm not sure I understand if `kafkaTarget` and `kafkaSource` are supposed to be URL schemes. If so, I don't see any tests that use these URL schemes.

And there's a another issue. URL schemes [are case-insensitive](https://www.rfc-editor.org/rfc/rfc3986#section-3.1).
> Although schemes are case-
   insensitive, the canonical form is lowercase and documents that
   specify schemes must do so with lowercase letters.  An implementation
   should accept uppercase letters as equivalent to lowercase in scheme
   names (e.g., allow "HTTP" as well as "http") for the sake of
   robustness but should only produce lowercase scheme names for
   consistency.

File: tests/system/feature_store/test_feature_store.py [hunk 4]
```diff
@@ -3130,8 +3127,41 @@ def test_onehot_with_array_values(self):
     @pytest.mark.skipif(
         not kafka_brokers, reason="MLRUN_SYSTEM_TESTS_KAFKA_BROKERS must be set"
     )
-    def test_kafka_target(self, kafka_consumer):
+    def test_kafka_target_datastore_profile(self, kafka_consumer):
```
Comment (gtopper) [inline tests/system/feature_store/test_feature_store.py:3130]: Just want to make sure that you ran it, since it doesn't run in CI.

**Review by gtopper — CHANGES_REQUESTED** [review]

File: mlrun/datastore/datastore.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (alxtkr77) [reply mlrun/datastore/datastore.py:54]: This isn't about URLs. When using a DS profile, the schema is stored within the profile itself and not the URL. The sequence of events is as follows:
- Establish a KafkaTarget DS profile named "kfk_target".
- Later, refer to a target using "d://kfk_target".
- The `ingest()` function invokes `get_or_create_store("d://kfk_target")`.
- Within the `get_or_create_store()` function, the target's schema is retrieved and `schema_to_store("kafkaTarget")` is then called.

File: mlrun/datastore/s3.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (alxtkr77) [reply mlrun/datastore/s3.py:121]: > Then better move it into its own module. See [DatabricksFileSystemDisableCache](https://github.com/mlrun/mlrun/blob/v1.5.0-rc10/mlrun/datastore/dbfs_store.py#L62) for an example.
> 
> And with a capital W :)

will move

[the remark this answers is not in the document]
Comment (alxtkr77) [reply mlrun/datastore/s3.py:121]: > Btw, are we now going to have to do this also for azure, gcs, dbfs, and everything else?

We probably will. But I am not familiar with those, so if we decide that they will have same functionality we should refactor

File: mlrun/api/db/sqldb/db.py [hunk 5]
```diff
@@ -4035,8 +4035,8 @@ def list_datastore_profiles(
         project = project or config.default_project
         query_results = self._query(session, DatastoreProfile, project=project)
         return [
-            self._transform_datastore_profile_model_to_schema(query)
-            for query in query_results
+            self._transform_datastore_profile_model_to_schema(datastore_record)
+            for datastore_record in query_results
```
[the remark this answers is not in the document]
Comment (alxtkr77) [reply mlrun/api/db/sqldb/db.py:4039]: > I think might as well rename `query_results` to `datastore_records`.

will do

File: mlrun/datastore/datastore_profile.py [hunk 6]
```diff
@@ -199,3 +275,30 @@ def datastore_profile_read(url):
...
+
+    profile = datastore_profile_read(url)
+    parsed_url = urlparse(url)
+    scheme = profile.type
+    # Add scheme as a password to the network location part
+    netloc = (
+        f"{parsed_url.username}:{scheme}@{parsed_url.netloc}"
+        if parsed_url.username
+        else f":{scheme}@{parsed_url.netloc}"
+    )
```
[the remark this answers is not in the document]
Comment (alxtkr77) [reply mlrun/datastore/datastore_profile.py:291]: will do

File: mlrun/datastore/datastore_profile.py [hunk 7]
```diff
@@ -199,3 +275,30 @@ def datastore_profile_read(url):
 
 def register_temporary_client_datastore_profile(profile: DatastoreProfile):
     TemporaryClientDatastoreProfiles().add(profile)
+
+
+def datastore_profile_embed_url_scheme(url):
+    from urllib.parse import urlparse, urlunparse
```
[the remark this answers is not in the document]
Comment (alxtkr77) [reply mlrun/datastore/datastore_profile.py:281]: will do

File: mlrun/datastore/datastore_profile.py [hunk 8]
```diff
@@ -66,6 +67,75 @@ class DatastoreProfileBasic(DatastoreProfile):
...
+        if self.kwargs_private:
+            attributes = merge(attributes, self.kwargs_private)
+
+        if isinstance(self.topics, str):
+            topics = [self.topics]
+        if isinstance(self.brokers, str):
+            brokers = [self.brokers]
+
+        attributes["brokers"] = brokers
+        attributes["topics"] = topics
```
[the remark this answers is not in the document]
Comment (alxtkr77) [reply mlrun/datastore/datastore_profile.py:113]: will fix

File: mlrun/datastore/datastore_profile.py [hunk 9]
```diff
@@ -66,6 +67,75 @@ class DatastoreProfileBasic(DatastoreProfile):
...
+class DatastoreProfileKafkaTarget(DatastoreProfile):
+    type: str = pydantic.Field("kafkaTarget")
+    _private_attributes = "kwargs_private"
+    bootstrap_servers: str
+    topic: str
+    kwargs_public: typing.Optional[typing.Dict]
+    kwargs_private: typing.Optional[typing.Dict]
+
+    def attributes(self):
+        result = {"bootstrap_servers": self.bootstrap_servers}
```
[the remark this answers is not in the document]
Comment (alxtkr77) [reply mlrun/datastore/datastore_profile.py:79]: will fix

File: tests/datastore/test_base.py [hunk 10]
```diff
@@ -102,6 +103,43 @@ def test_kafka_source_with_attributes():
     }
 
 
+def test_kafka_source_with_attributes_as_ds_profile():
+    from unittest.mock import Mock
```
[the remark this answers is not in the document]
Comment (alxtkr77) [reply tests/datastore/test_base.py:107]: will fix

File: mlrun/datastore/datastore_profile.py [hunk 11]
```diff
@@ -66,6 +67,75 @@ class DatastoreProfileBasic(DatastoreProfile):
...
+        attributes["initial_offset"] = self.initial_offset
+        if self.partitions is not None:
+            attributes["partitions"] = self.partitions
+        sasl = attributes.pop("sasl", {})
+        if self.sasl_user and self.sasl_pass:
+            sasl["enabled"] = True
+            sasl["user"] = self.sasl_user
+            sasl["password"] = self.sasl_pass
+        if sasl:
+            attributes["sasl"] = sasl
```
[the remark this answers is not in the document]
Comment (alxtkr77) [reply mlrun/datastore/datastore_profile.py:124]: > Whaat about this one?
> Also, I'm not really following why kafkaTarget scheme should get us to return a FileStore?
As a matter of fact it does. The kafkaTarget path is not an url - it looks like a file path and returns FileStore (why - I don't know. This is a question to the kafka drive authors :). Once I added ds://.../ this started to fall. Also see my previous response https://github.com/mlrun/mlrun/pull/4088#discussion_r1323996582

[the remark this answers is not in the document]
Comment (alxtkr77) [reply mlrun/datastore/datastore_profile.py:124]: > 

will fix

File: mlrun/datastore/s3.py [hunk 12]
```diff
@@ -104,30 +117,47 @@ def get_filesystem(self, silent=False):
...
+                    parsed_url = urlparse(url)
+                    url = parsed_url.path[1:]
+                return super()._strip_protocol(url)
+
+        self._filesystem = S3withDS(**self.get_storage_options())
         return self._filesystem
 
     def get_storage_options(self):
+        if self.kind == "ds":
+            datastore_profile = datastore_profile_read(self.name)
```
[the remark this answers is not in the document]
Comment (alxtkr77) [reply mlrun/datastore/s3.py:134]: will add a comment

File: mlrun/datastore/s3.py [hunk 13]
```diff
@@ -137,35 +167,46 @@ def get_storage_options(self):
 
         return storage_options
 
+    def get_bucket_and_key(self, key):
+        path = self._join(key)[1:]
+        if self.endpoint != "":
```
[the remark this answers is not in the document]
Comment (alxtkr77) [reply mlrun/datastore/s3.py:172]: will fix

File: tests/datastore/test_base.py [hunk 14]
```diff
@@ -102,6 +103,43 @@ def test_kafka_source_with_attributes():
     }
 
 
+def test_kafka_source_with_attributes_as_ds_profile():
```
[the remark this answers is not in the document]
Comment (alxtkr77) [reply tests/datastore/test_base.py:106]: the sasl thing is covered in the unit tests. As for the list of brokers - will add a test

File: tests/integration/aws_s3/test_aws_s3.py [hunk 15]
```diff
@@ -49,9 +54,31 @@ def aws_s3_configured(extra_params=None):
...
+    ):
+        res = {}
+        res["bucket_path"] = prefix + bucket_name
+        res["object_path"] = object_dir + "/" + object_file
+        res["df_path"] = object_dir + "/" + csv_file
+
+        res["object_url"] = res["bucket_path"] + "/" + res["object_path"]
+        res["df_url"] = res["bucket_path"] + "/" + res["df_path"]
+        res["blob_url"] = res["object_url"] + ".blob"
+        res["parquet_url"] = res["object_url"] + ".parquet"
```
[the remark this answers is not in the document]
Comment (alxtkr77) [reply tests/integration/aws_s3/test_aws_s3.py:76]: legacy... will fix

File: tests/integration/aws_s3/test_aws_s3.py [hunk 16]
```diff
@@ -49,9 +54,31 @@ def aws_s3_configured(extra_params=None):
     return True
 
 
+@TestMLRunSystem.skip_test_if_env_not_configured
 @pytest.mark.skipif(not aws_s3_configured(), reason="AWS S3 parameters not configured")
-class TestAwsS3:
+@pytest.mark.parametrize("use_datastore_profile", [False, True])
+@pytest.mark.enterprise
+class TestAwsS3(TestMLRunSystem):
```
[the remark this answers is not in the document]
Comment (alxtkr77) [reply tests/integration/aws_s3/test_aws_s3.py:61]: will move

File: tests/integration/aws_s3/test_aws_s3.py [hunk 17]
```diff
@@ -60,20 +87,31 @@ def setup_method(self, method):
...
+        )
+        self.s3["ds"] = self._make_target_names(
+            "ds://s3ds_profile/", self._bucket_name, object_dir, object_file, csv_file
+        )
+
+        logger.info(f'DS Object URL: {self.s3["ds"]["object_url"]}')
+        logger.info(f'S3 Object URL: {self.s3["s3"]["object_url"]}')
+
+        project = mlrun.get_or_create_project("default", "./")
+        profile = DatastoreProfileS3(
```
[the remark this answers is not in the document]
Comment (alxtkr77) [reply tests/integration/aws_s3/test_aws_s3.py:102]: we don't. I just added the ds profile creation here. Added @pytest.mark.parametrize("use_datastore_profile", [False, True]) parameter

File: mlrun/datastore/s3.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (assaf758) [reply mlrun/datastore/s3.py:121]: yes, plan is to complete the support in 1.6 ([ML-4343](https://jira.iguazeng.com/browse/ML-4343))

File: mlrun/datastore/datastore.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (gtopper) [reply mlrun/datastore/datastore.py:54]: See
```
schema = parsed_url.scheme.lower()
```
in `parse_url()`.
It clearly gets the "schema" out of the URL **scheme**.

I'm not sure how it works when the scheme is shifted to lower case, and then later compared against `"kafkaTarget"` in `schema_to_store()`.

[the remark this answers is not in the document]
Comment (gtopper) [reply mlrun/datastore/datastore.py:54]: And anything on `+ list(string.ascii_lowercase)`? It looks like a (preexisting) bug.

File: tests/datastore/test_base.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (gtopper) [reply tests/datastore/test_base.py:106]: Taking sasl from kwargs is covered? The test I see override the sasl in kwargs with `attributes["sasl"]`.

File: tests/integration/aws_s3/test_aws_s3.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (gtopper) [reply tests/integration/aws_s3/test_aws_s3.py:61]: Please don't move existing integration test into system though.

File: tests/integration/aws_s3/test_aws_s3.py [hunk 18]
```diff
@@ -60,20 +87,31 @@ def setup_method(self, method):
...
+            "s3://", self._bucket_name, object_dir, object_file, csv_file
+        )
+        self.s3["ds"] = self._make_target_names(
+            "ds://s3ds_profile/", self._bucket_name, object_dir, object_file, csv_file
+        )
+
+        logger.info(f'DS Object URL: {self.s3["ds"]["object_url"]}')
+        logger.info(f'S3 Object URL: {self.s3["s3"]["object_url"]}')
+
+        project = mlrun.get_or_create_project("default", "./")
```
Comment (gtopper) [inline tests/integration/aws_s3/test_aws_s3.py:101]: Shouldn't this be
```suggestion
        project = mlrun.get_or_create_project(self.project_name))
```
?

Comment (gtopper) [reply tests/integration/aws_s3/test_aws_s3.py:101]: Or
```suggestion
        project = mlrun.get_or_create_project(self.project_name, "./")
```

File: tests/integration/aws_s3/test_aws_s3.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (alxtkr77) [reply tests/integration/aws_s3/test_aws_s3.py:61]: Actually, will fix. Now we have register_temporary_client_datastore_profile() , so no need to have a project to test datastore profiles

File: tests/integration/aws_s3/test_aws_s3.py (hunk 18 above)
Comment (alxtkr77) [reply tests/integration/aws_s3/test_aws_s3.py:101]: will remove

File: tests/system/test_aws_s3.py [hunk 19]
```diff
@@ -0,0 +1,104 @@
...
+
+test_environment = TestMLRunSystem._get_env_from_file()
+aws_keys = ["AWS_ACCESS_KEY_ID", "AWS_SECRET_ACCESS_KEY", "AWS_BUCKET_NAME"]
+
+
+@TestMLRunSystem.skip_test_if_env_not_configured
+@pytest.mark.skipif(
+    not all(test_environment[key] for key in aws_keys),
+    reason=f"one of {aws_keys} keys is not set, skipping until testing against real aws",
+)
```
Comment (gtopper) [inline tests/system/test_aws_s3.py:42]: This skipif doesn't work as it should. I get
```
KeyError: 'AWS_ACCESS_KEY_ID'
```
when I try to run this suite without setting the env vars.

Also, doing it like this will result in the user not knowing which variable is missing. This matters if we get this failure in CI and we can't just check it in the debugger.

I suggest:
```suggestion

@TestMLRunSystem.skip_test_if_env_not_configured
@pytest.mark.skipif(
    not test_environment.get("AWS_ACCESS_KEY_ID"),
    reason=f"AWS_ACCESS_KEY_ID is not set",
)
@pytest.mark.skipif(
    not test_environment.get("AWS_SECRET_ACCESS_KEY"),
    reason=f"AWS_SECRET_ACCESS_KEY is not set",
)
@pytest.mark.skipif(
    not test_environment.get("AWS_BUCKET_NAME"),
    reason=f"AWS_BUCKET_NAME is not set",
)
```

File: tests/system/test_aws_s3.py [hunk 20]
```diff
@@ -0,0 +1,104 @@
...
+@pytest.mark.parametrize("use_datastore_profile", [True, False])
+class TestAwsS3(TestMLRunSystem):
+    def custom_setup(self):
+        pass
+
```
Comment (gtopper) [inline tests/system/test_aws_s3.py:47]: Accidental leftover:
```suggestion
```

File: tests/system/test_aws_s3.py [hunk 21]
```diff
@@ -0,0 +1,104 @@
...
+    DatastoreProfileS3,
+    register_temporary_client_datastore_profile,
+)
+from mlrun.datastore.targets import ParquetTarget
+from mlrun.utils import logger
+from tests.system.base import TestMLRunSystem
+
+# Used to test dataframe functionality (will be saved as csv)
+test_df_string = "col1,col2,col3\n1,2,3"
+
```
Comment (gtopper) [inline tests/system/test_aws_s3.py:33]: Actually it's not used at all
```suggestion
```

File: tests/system/test_aws_s3.py [hunk 22]
```diff
@@ -0,0 +1,104 @@
...
+        object_file = f"file_{random.randint(0, 1000)}.txt"
+        csv_file = f"file_{random.randint(0,1000)}.csv"
+
+        self.s3 = {}
+        self.s3["s3"] = self._make_target_names(
+            "s3://", self._bucket_name, object_dir, object_file, csv_file
+        )
+        self.s3["ds"] = self._make_target_names(
+            "ds://s3ds_profile/", self._bucket_name, object_dir, object_file, csv_file
+        )
```
Comment (gtopper) [inline tests/system/test_aws_s3.py:82]: Screenshot 2023-09-18 at 13 49 38
Not a linter-enforced style rule, but I think they have a point.

File: tests/system/feature_store/test_feature_store.py [hunk 23]
```diff
@@ -125,7 +126,10 @@ def _generate_random_name():
     return random_name
 
 
-kafka_brokers = os.getenv("MLRUN_SYSTEM_TESTS_KAFKA_BROKERS")
+test_environment = TestMLRunSystem._get_env_from_file()
+kafka_brokers = test_environment["MLRUN_SYSTEM_TESTS_KAFKA_BROKERS"] or os.getenv(
```
Comment (gtopper) [inline tests/system/feature_store/test_feature_store.py:130]:
```suggestion
kafka_brokers = test_environment.get("MLRUN_SYSTEM_TESTS_KAFKA_BROKERS") or os.getenv(
```
or it will raise an error and never hit the fallback.

File: tests/system/feature_store/test_feature_store.py [hunk 24]
```diff
@@ -260,9 +264,7 @@ def _get_offline_vector(
         )
         resp = fstore.get_offline_features(
             vector,
-            entity_rows=trades.set_index(
-                "ticker"
-            ),  # test when the relation keys are indexes.
```
Comment (gtopper) [inline tests/system/feature_store/test_feature_store.py:265]: Why is this removed? It doesn't seem related to this PR.

File: tests/integration/aws_s3/test_aws_s3.py [hunk 25]
```diff
@@ -50,7 +54,29 @@ def aws_s3_configured(extra_params=None):
 
 
 @pytest.mark.skipif(not aws_s3_configured(), reason="AWS S3 parameters not configured")
+@pytest.mark.parametrize("use_datastore_profile", [False, True])
 class TestAwsS3:
+    def custom_setup(self):
+        pass
+
```
Comment (gtopper) [inline tests/integration/aws_s3/test_aws_s3.py:61]: Seems like an accidental leftover
```suggestion
```

File: tests/system/env-template.yml [hunk 26]
```diff
@@ -50,3 +50,9 @@ MLRUN_SYSTEM_TESTS_SLACK_WEBHOOK_URL:
 
 # sql db path string for test_sql_db - e.g. mysql+pymysql://<username>:<password>@<host>:<port>/<db_name>
 MLRUN_SQL__URL:
+
+# AWS bucket name
```
Comment (gtopper) [inline tests/system/env-template.yml:54]: Self explanatory
```suggestion
```

File: tests/system/env-template.yml [hunk 27]
```diff
@@ -50,3 +50,9 @@ MLRUN_SYSTEM_TESTS_SLACK_WEBHOOK_URL:
...
+AWS_BUCKET_NAME: 
+# Access key parameters as created by AWS IAM secret key mechanism
```
Comment (gtopper) [inline tests/system/env-template.yml:56]: I don't think we need to explain it here either
```suggestion
```

File: mlrun/datastore/datastore.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (alxtkr77) [reply mlrun/datastore/datastore.py:54]: > And anything on `+ list(string.ascii_lowercase)`? It looks like a (preexisting) bug.

This is to support windows full path  including drive letter , something like `c:/some/path`

File: tests/datastore/test_base.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (alxtkr77) [reply tests/datastore/test_base.py:107]: > Taking sasl from kwargs is covered? The test I see override the sasl in kwargs with `attributes["sasl"]`.
Right. Check out the last commit - I added more tests

[the remark this answers is not in the document]
Comment (alxtkr77) [reply tests/datastore/test_base.py:107]: Right. Check out the last commit - I added more tests

File: tests/system/test_aws_s3.py (hunk 19 above)
Comment (alxtkr77) [reply tests/system/test_aws_s3.py:42]: Will do

File: tests/system/test_aws_s3.py (hunk 20 above)
Comment (alxtkr77) [reply tests/system/test_aws_s3.py:47]: will do

File: tests/system/test_aws_s3.py (hunk 21 above)
Comment (alxtkr77) [reply tests/system/test_aws_s3.py:33]: will do

File: tests/system/test_aws_s3.py (hunk 22 above)
Comment (alxtkr77) [reply tests/system/test_aws_s3.py:82]: will do

Comment (alxtkr77) [reply tests/system/test_aws_s3.py:82]: will do

File: tests/system/feature_store/test_feature_store.py (hunk 23 above)
Comment (alxtkr77) [reply tests/system/feature_store/test_feature_store.py:130]: will fix

File: tests/system/feature_store/test_feature_store.py (hunk 4 above)
Comment (alxtkr77) [reply tests/system/feature_store/test_feature_store.py:3133]: I did

File: tests/system/feature_store/test_feature_store.py (hunk 24 above)
Comment (alxtkr77) [reply tests/system/feature_store/test_feature_store.py:265]: right - will fix

File: tests/integration/aws_s3/test_aws_s3.py (hunk 25 above)
Comment (alxtkr77) [reply tests/integration/aws_s3/test_aws_s3.py:61]: will fix

File: tests/system/env-template.yml (hunk 26 above)
Comment (alxtkr77) [reply tests/system/env-template.yml:54]: will fix

File: tests/system/env-template.yml (hunk 27 above)
Comment (alxtkr77) [reply tests/system/env-template.yml:56]: will fix

File: mlrun/datastore/datastore.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (alxtkr77) [reply mlrun/datastore/datastore.py:54]: Somehow I cannot reply to 'https://github.com/mlrun/mlrun/pull/4088#discussion_r1322396996', so I do it here.
Why kafkaTarget scheme should get us to return a FileStore I don't know - the question goes to the authors of kafka target :). Prior the datastore profile kafka path was not an URL but just `path/to/kafka` that has no schema and that bounced to FileStore.

[the remark this answers is not in the document]
Comment (alxtkr77) [reply mlrun/datastore/datastore.py:54]: I will change kafkaTarget to kafka_target - maybe in the future we will want to make it a part of URL

[the remark this answers is not in the document]
Comment (gtopper) [reply mlrun/datastore/datastore.py]: > This is to support windows full path including drive letter , something like c:/some/path
Can you please add a comment? I thought it might be the case, but it also looks like it might be a bug.

File: mlrun/datastore/datastore.py [hunk 28]
```diff
@@ -51,7 +51,10 @@ def parse_url(url):
 
 def schema_to_store(schema):
     # import store classes inside to enable making their dependencies optional (package extras)
-    if not schema or schema in ["file"] + list(string.ascii_lowercase):
+
+    # The expression `list(string.ascii_lowercase)` generates a list of lowercase alphabets,
+    # which corresponds to drive letters in Windows file paths such as `C:/Windows/path`.
```
Comment (gtopper) [inline mlrun/datastore/datastore.py:56]: Ok, now you're making me worry... Should it be lowercase or uppercase? 😬

Comment (alxtkr77) [reply mlrun/datastore/datastore.py:56]: Like you said, the schema (or in this case drive letter) is always parsed as a low case. So when C:/Windows path is parsed, the schema will be 'c' and not 'C'. No worries :)

Comment (gtopper) [reply mlrun/datastore/datastore.py:56]: I just found the discrepancy a bit confusing.

**Review by gtopper — APPROVED** [review]
